metrics/stackdriver_exporter.go

```diff
@@ -40,8 +40,10 @@ const (
 	// defaultCustomMetricSubDomain is the default subdomain to use for unsupported metrics by monitored resource types.
 	// See: https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors#MetricDescriptor
 	defaultCustomMetricSubDomain = "knative.dev"
-	// secretNamespaceDefault is the namespace to search for a k8s Secret to pass to Stackdriver client to authenticate with Stackdriver.
-	secretNamespaceDefault = "default"
+	// DefaultSecretNamespace is the default namespace to search for a k8s Secret to pass to Stackdriver client to authenticate with Stackdriver.
+	DefaultSecretNamespace = "default"
```

Comment:
Should this be system.Namespace()?
https://github.com/knative/pkg/blob/50398a29a78804bf5e9e64d937849c2747b41dd6/system/names.go#L30

Reply:
Since we're in knative/pkg which should be pretty generic, I don't think panicking is a good option if that environment variable isn't set.

Reply:
What about defaulting to the empty string and then lazily reading system.Namespace() if this value is empty?

Reply:
That still leaves the possibility of the metrics pkg causing a panic unnecessarily if that env is missing. Also, I believe empty string is a valid namespace string that conflates to "default" for kubeclient APIs. Consuming packages should be allowed to set the secretNamespace to empty string. Consuming packages can also set the namespace to system.Namespace().

Is the concern here that metrics pkg consumers will accidentally trigger a Secret read in the default namespace without being aware of it?

Reply:
Also, we cannot allow reading an env variable because that is user-configurable which will be a security issue for knative/serving when it starts to propagate the Secret specified here to queue-proxy.

Reply:
ok I've added a metrics pkg internal bool for "useStackdriverSecretEnabled" that prevents any attempts to read the Secret until SetStackdriverSecretLocation is called.

More details in the new description for SetStackdriverSecretLocation.

This means that the DefaultSecretName & DefaultSecretNamespace values are never used unless explicitly re-specified by the consuming package. (They are still provided so that packages don't all make up a different Secret name).